injectproxy/routes.go

```diff
@@ -118,27 +120,65 @@ func (r *routes) noop(w http.ResponseWriter, req *http.Request) {
 }
 
 func (r *routes) query(w http.ResponseWriter, req *http.Request) {
-	expr, err := parser.ParseExpr(req.FormValue("query"))
+	// The `query` can come in the URL query string and/or the POST body.
+	// For this reason, we need to try to enforcing in both places.
+	// Note: a POST request may include some values in the URL query string
+	// and others in the body. If both locations include a `query`, then
+	// enforce in both places.
+	q, f1, err := enforceValues(req.Context(), r.label, req.URL.Query())
 	if err != nil {
 		return
 	}
+	req.URL.RawQuery = q
+
+	var f2 bool
+	// Enforce the query in the POST body if needed.
+	if req.Method == http.MethodPost {
```

Comment:
I would always do that. This is because GET can have body as well (: Wonder if that was not a leak before this impl then (I think Prom takes form no matter what, also for GETs)

https://stackoverflow.com/questions/978061/http-get-with-request-body

Reply:
Also I wonder if we can just remove the upper one and always parse Form. This is because Form takes care of query as well: 

```
// ParseForm populates r.Form and r.PostForm.
//
// For all requests, ParseForm parses the raw query from the URL and updates
// r.Form.
//
// For POST, PUT, and PATCH requests, it also reads the request body, parses it
// as a form and puts the results into both r.PostForm and r.Form. Request body
// parameters take precedence over URL query string values in r.Form.
//
// If the request Body's size has not already been limited by MaxBytesReader,
// the size is capped at 10MB.
//
// For other HTTP methods, or when the Content-Type is not
// application/x-www-form-urlencoded, the request Body is not read, and
// r.PostForm is initialized to a non-nil, empty value.
//
// ParseMultipartForm calls ParseForm automatically.
// ParseForm is idempotent.
```

Reply:
No we don't want to remove this. Again, values can come from EITHER one of the places OR both. There are some really funky things that can happen if someone puts a query into BOTH the query string AND the form body. Imagine that tomorrow Prometheus stops using FormValue and only reads the query from one of them, then we would only override the form body but keep the unescaped query in the query string. This would very easily lead to a vulnerability where the unescaped query in the URL is executed. Or imagine that a proxy in the middle of the promlabelproxy and Prometheus only uses the url query string. Same thing would happen. We need to be defensive and escape both queries. We cannot do this if we use FormValue, so we need to explicitly look for queries in both places

Reply:
> This would very easily lead to a vulnerability where the unescaped query in the URL is executed

Agree. This is why vote for clearing out query param completely.

Reply:
Tested this thoroughly and it is not a vulnerability today due to shear luck because of the form parsing logic in go.

> but rather vote for clearing out query param completely on POST.

Why? The user sent both so let's escape both. We are dropping user data otherwise. Essentially are codifying the implementation logic of Go's form parsing as business logic of this proxy. They are two different domains and should not be mixed. If Go preferred query parameters over form parameters what would we do? And what if we used Python instead? etc

Reply:
> Tested this thoroughly and it is not a vulnerability today due to shear luck because of the form parsing logic in go.

Thanks!

Added test case as well in https://github.com/prometheus-community/prom-label-proxy/pull/53/commits/c9da73397b6c60a3326bd6cf06b99261dadc27e5